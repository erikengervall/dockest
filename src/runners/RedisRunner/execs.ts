import execa from 'execa'
import { defaultDockerComposeRunOpts } from '../../constants'
import { DockestError } from '../../errors'
import { RunnerLogger } from '../../loggers'
import { acquireConnection, getContainerId, sleep, teardownSingle } from '../utils'
import { RedisRunnerConfig } from './index'

interface Exec {
  start: (runnerConfig: RedisRunnerConfig, runnerKey: string) => Promise<string>
  checkHealth: (
    runnerConfig: RedisRunnerConfig,
    containerId: string,
    runnerKey: string
  ) => Promise<void>
  teardown: (containerId: string, runnerKey: string) => Promise<void>
}

class RedisExec implements Exec {
  private static instance: RedisExec

  constructor() {
    return RedisExec.instance || (RedisExec.instance = this)
  }

  public start = async (runnerConfig: RedisRunnerConfig, runnerKey: string) => {
    RunnerLogger.startContainer(runnerKey)

    const { port, service, password } = runnerConfig

    let containerId = await getContainerId(service)
    if (!containerId) {
      const portMapping = `--publish ${port}:6379`
      const auth = !!password ? `--requirepass ${password}` : ''
      const cmd = `docker-compose run \
                    ${defaultDockerComposeRunOpts} \
                    ${portMapping} \
                    ${service} \
                    ${auth}`
      RunnerLogger.shellCmd(cmd)
      await execa.shell(cmd)
    }
    containerId = await getContainerId(service)

    RunnerLogger.startContainerSuccess(runnerKey)

    return containerId
  }

  public checkHealth = async (
    runnerConfig: RedisRunnerConfig,
    containerId: string,
    runnerKey: string
  ) => {
    RunnerLogger.checkHealth(runnerKey)

    await this.checkResponsiveness(runnerConfig, containerId, runnerKey)
    await this.checkConnection(runnerConfig, runnerKey)

    RunnerLogger.checkHealthSuccess(runnerKey)
  }

  public teardown = async (containerId: string, runnerKey: string) =>
    teardownSingle(containerId, runnerKey)

  private checkResponsiveness = async (
    runnerConfig: RedisRunnerConfig,
    containerId: string,
    runnerKey: string
  ) => {
    const { responsivenessTimeout, host: runnerHost, password: runnerPassword } = runnerConfig

    const recurse = async (responsivenessTimeout: number): Promise<void> => {
      RunnerLogger.checkResponsiveness(runnerKey, responsivenessTimeout)

      if (responsivenessTimeout <= 0) {
        throw new DockestError(`Redis responsiveness timed out`)
      }

      try {
        const host = `-h ${runnerHost}`
        const port = `-p 6379`
        const password = runnerPassword ? `-a ${runnerPassword}` : ''
        const command = `PING`
        const redisCliOpts = `${host} \
                              ${port} \
                              ${password} \
                              ${command}`
        const cmd = `docker exec ${containerId} redis-cli ${redisCliOpts}`
        RunnerLogger.shellCmd(cmd)
        const { stdout: result } = await execa.shell(cmd)
        if (result !== 'PONG') {
          throw new Error('PING did not recieve a PONG')
        }

        RunnerLogger.checkResponsivenessSuccess(runnerKey)
      } catch (error) {
        responsivenessTimeout--

        await sleep(1000)
        await recurse(responsivenessTimeout)
      }
    }

    await recurse(responsivenessTimeout)
  }

  private checkConnection = async (
    runnerConfig: RedisRunnerConfig,
    runnerKey: string
  ): Promise<void> => {
    const { connectionTimeout, host, port } = runnerConfig

    const recurse = async (connectionTimeout: number) => {
      RunnerLogger.checkConnection(runnerKey, connectionTimeout)

      if (connectionTimeout <= 0) {
        throw new DockestError(`Redis connection timed out`)
      }

      try {
        await acquireConnection(port, host)

        RunnerLogger.checkConnectionSuccess(runnerKey)
      } catch (error) {
        connectionTimeout--

        await sleep(1000)
        await recurse(connectionTimeout)
      }
    }

    await recurse(connectionTimeout)
  }
}

export default RedisExec
