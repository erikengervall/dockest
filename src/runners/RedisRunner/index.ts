import Logger from '../../Logger'
import validateConfig from '../../utils/validateConfig'
import validateTypes from '../../utils/validateTypes'
import { BaseRunner, GetComposeService, SharedRequiredConfigProps, SharedDefaultableConfigProps } from '../@types'
import { SHARED_DEFAULT_CONFIG_PROPS } from '../constants'
import { defaultGetComposeService } from '../composeFileHelper'

interface RequiredConfigProps extends SharedRequiredConfigProps {}
interface DefaultableConfigProps extends SharedDefaultableConfigProps {
  password: string
  responsivenessTimeout: number
}
interface RedisRunnerConfig extends RequiredConfigProps, DefaultableConfigProps {}

const DEFAULT_PORT = '6379'
const DEFAULT_CONFIG: DefaultableConfigProps = {
  ...SHARED_DEFAULT_CONFIG_PROPS,
  password: '',
  ports: {
    [DEFAULT_PORT]: DEFAULT_PORT,
  },
  responsivenessTimeout: SHARED_DEFAULT_CONFIG_PROPS.responsivenessTimeout,
}

class RedisRunner implements BaseRunner {
  public static DEFAULT_HOST = SHARED_DEFAULT_CONFIG_PROPS.host
  public static DEFAULT_PORT = DEFAULT_PORT
  public containerId = ''
  public initializer = ''
  public runnerConfig: RedisRunnerConfig
  public logger: Logger

  public constructor(config: RequiredConfigProps & Partial<DefaultableConfigProps>) {
    this.runnerConfig = {
      ...DEFAULT_CONFIG,
      ...config,
    }
    this.logger = new Logger(this)

    const schema: { [key in keyof RequiredConfigProps]: any } = {
      service: validateTypes.isString,
    }
    validateConfig(schema, this.runnerConfig)
  }

  public getComposeService: GetComposeService = () => defaultGetComposeService(this.runnerConfig)

  public createResponsivenessCheckCmd = () => {
    const { host: runnerHost, password: runnerPassword } = this.runnerConfig
    const containerId = this.containerId

    // FIXME: Should `-p` really be DEFAULT_PORT or runnerConfig's port?
    const redisCliPingOpts = ` \
                            -h ${runnerHost} \
                            -p ${DEFAULT_PORT} \
                            ${!!runnerPassword ? `-a ${runnerPassword}` : ''} \
                            PING \
                          `
    const command = ` \
                      docker exec ${containerId} redis-cli ${redisCliPingOpts} \
                    `

    return command
  }
}

export { RedisRunnerConfig }
export default RedisRunner
