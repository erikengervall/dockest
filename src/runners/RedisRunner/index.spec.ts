import RedisRunner from './index'

const config = {
  service: '_',
}
const RedisRunner1 = new RedisRunner(config)
const RedisRunner2 = new RedisRunner(config)

describe('RedisRunner', () => {
  it('should create unique instances', () => {
    expect(RedisRunner1).not.toBe(RedisRunner2)
  })
})
