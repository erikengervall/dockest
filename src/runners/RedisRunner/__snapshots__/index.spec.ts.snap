// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RedisRunner should create unique instances 1`] = `
RedisRunner {
  "containerId": "",
  "createResponsivenessCheckCmd": [Function],
  "getComposeService": [Function],
  "initializer": "",
  "isBridgeNetworkMode": false,
  "logger": Logger {
    "debug": [Function],
    "error": [Function],
    "info": [Function],
    "runnerService": "r1",
    "runnerSymbol": "ðŸ¦‡ ",
    "setRunnerSymbol": [Function],
    "warn": [Function],
  },
  "runnerConfig": Object {
    "build": "",
    "commands": Array [],
    "connectionTimeout": 5,
    "dependsOn": Array [],
    "host": "localhost",
    "image": "some/image:123",
    "networks": undefined,
    "password": "",
    "ports": Array [
      Object {
        "published": 6379,
        "target": 6379,
      },
    ],
    "props": Object {},
    "responsivenessTimeout": 30,
    "service": "r1",
  },
}
`;

exports[`RedisRunner should create unique instances 2`] = `
RedisRunner {
  "containerId": "",
  "createResponsivenessCheckCmd": [Function],
  "getComposeService": [Function],
  "initializer": "",
  "isBridgeNetworkMode": false,
  "logger": Logger {
    "debug": [Function],
    "error": [Function],
    "info": [Function],
    "runnerService": "r2",
    "runnerSymbol": "ðŸ¦‡ ",
    "setRunnerSymbol": [Function],
    "warn": [Function],
  },
  "runnerConfig": Object {
    "build": "",
    "commands": Array [],
    "connectionTimeout": 5,
    "dependsOn": Array [],
    "host": "localhost",
    "image": "some/image:123",
    "networks": undefined,
    "password": "",
    "ports": Array [
      Object {
        "published": 6379,
        "target": 6379,
      },
    ],
    "props": Object {},
    "responsivenessTimeout": 30,
    "service": "r2",
  },
}
`;
